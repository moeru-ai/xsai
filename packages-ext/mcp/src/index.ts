export * from './utils/get-tools'
