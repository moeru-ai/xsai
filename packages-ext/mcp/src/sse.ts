import type { SSEClientTransportOptions } from '@modelcontextprotocol/sdk/client/sse.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js'

export type CreateSSEClientsOptions = Record<string, SSEClientTransportOptions & {
  url: URL
}>

/** @experimental */
export const createSSEClients = async (mcpServers: CreateSSEClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, { url, ...options }]) => {
          const transport = new SSEClientTransport(url, options)
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
