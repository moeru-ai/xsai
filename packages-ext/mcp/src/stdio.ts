import type { StdioServerParameters } from '@modelcontextprotocol/sdk/client/stdio.js'

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js'
import { stderr } from 'node:process'

export type CreateStdioClientsOptions = Record<string, StdioServerParameters>

/** @experimental */
export const createStdioClients = async (mcpServers: CreateStdioClientsOptions, version: string = '1.0.0'): Promise<Record<string, Client>> =>
  Object.fromEntries(
    await Promise.all(
      Object.entries(mcpServers)
        .map(async ([serverName, server]) => {
          const transport = new StdioClientTransport({ ...server, stderr })
          const client = new Client({ name: serverName, version })
          await client.connect(transport)

          return [serverName, client] as const
        }),
    ),
  )
