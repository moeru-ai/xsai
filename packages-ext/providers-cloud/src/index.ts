export * from './providers'

// TODO: providers list
