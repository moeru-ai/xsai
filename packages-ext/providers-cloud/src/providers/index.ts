export { createDeepSeek } from './deepseek'
export { createFireworksAI } from './fireworks-ai'
export { createGoogleGenerativeAI } from './google-generative-ai'
export { createGroq } from './groq'
export { createMinimax } from './minimax'
export { createMistralAI } from './mistral-ai'
export { createMoonshot } from './moonshot'
export { createNovita } from './novita'
export { createOpenAI } from './openai'
export { createOpenRouter } from './openrouter'
export { createPerplexity } from './perplexity'
export { createQwen } from './qwen'
export { createSiliconFlow } from './silicon-flow'
export { createStepfun } from './stepfun'
export { createTencentHunyuan } from './tencent-hunyuan'
export { createTogetherAI } from './together-ai'
export { createWorkersAI } from './workers-ai'
export { createXAI } from './xai'
export { createZhipu } from './zhipu'
