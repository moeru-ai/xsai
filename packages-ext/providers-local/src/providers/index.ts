export { createClaude2OpenAI } from './claude2openai'
export { createOllama } from './ollama'
export { createUnSpeech } from './unspeech'
export { createUnElevenLabs } from './unspeech/elevenlabs'
export { createUnMicrosoft } from './unspeech/microsoft'
