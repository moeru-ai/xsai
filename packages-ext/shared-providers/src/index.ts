export type * from './types/create-provider-options'
export type * from './types/providers'
export * from './utils/create'
export * from './utils/merge'
