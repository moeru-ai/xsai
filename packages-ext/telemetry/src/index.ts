export { generateText } from './utils/generate-text'
export { streamText } from './utils/stream-text'
