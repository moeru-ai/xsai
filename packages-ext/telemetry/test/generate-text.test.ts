import type { ReadableSpan } from '@opentelemetry/sdk-trace-base'

import { InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node'
import { generateText as aiGenerateText } from 'ai'
import { ollama } from 'ollama-ai-provider-v2'
import { describe, expect, it } from 'vitest'

import { generateText } from '../src'

describe.sequential('generateText', () => {
  const memoryExporter = new InMemorySpanExporter()
  const tracerProvider = new NodeTracerProvider({
    spanProcessors: [new SimpleSpanProcessor(memoryExporter)],
  })
  tracerProvider.register()

  it('basic', async () => {
    const { text } = await generateText({
      baseURL: 'http://localhost:11434/v1',
      messages: [{
        content: 'Why is the sky blue?',
        role: 'user',
      }],
      model: 'qwen3:0.6b',
      seed: 114514,
    })

    const spans = memoryExporter.getFinishedSpans()

    expect(text).toMatchSnapshot()
    expect(spans).toMatchSnapshot()
  })

  it('basic/ai', async () => {
    const { text } = await aiGenerateText({
      experimental_telemetry: { isEnabled: true },
      model: ollama('qwen3:0.6b'),
      prompt: 'Why is the sky blue?',
      seed: 114514,
    })

    const spans = memoryExporter.getFinishedSpans().slice(2)

    expect(text).toMatchSnapshot()
    expect(spans).toMatchSnapshot()
  })

  it('basic/compare', () => {
    const extractAttributes = (span: ReadableSpan) => Object.keys(span.attributes)
      // eslint-disable-next-line sonarjs/no-nested-functions
      .filter(key => ['ai.settings'].every(prefix => !key.startsWith(prefix)))
      .sort((a, b) => a.localeCompare(b))

    const spans = memoryExporter.getFinishedSpans()
    const xsai = spans.slice(0, 2).map(extractAttributes)
    const ai = spans.slice(2).map(extractAttributes)

    xsai.forEach((attributes, i) => {
      expect(attributes).toStrictEqual(ai[i])
    })
  })
})
