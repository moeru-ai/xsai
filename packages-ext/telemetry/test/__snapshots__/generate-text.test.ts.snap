// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateText > basic 1`] = `
"<think>
Okay, so the user is asking, "Why is the sky blue?" I need to explain why the sky appears blue. I remember from my studies that light is bending when it passes through the atmosphere, which is a key part of the rainbow effect. But I should make sure I'm explaining this clearly. First, let's start with the basic fact: when sunlight reaches Earth's atmosphere, it's refracted. The atmosphere has a mixture of gases like nitrogen and oxygen, and water vapor. So, the light is scattered by those gases, making the sky appear blue. Oh right, the Earth's atmosphere is much thinner than the atmosphere above the ocean, which is why the sky is usually visible without a sunscreen. But maybe the user is interested in how to explain this, not just the basic reason. Also, I should mention the color wavelength of sunlight, which is in the range of blue. So, maybe break it down into parts: refraction, scattering by gases, and the fact that the colors of the sky are due to that refraction. Let me check if I'm missing anything here, like any misconceptions, like maybe the user thinks it's a color that's always there, but the answer is based on why the color is blue. Make sure to use simple terms and avoid technical jargon. Also, perhaps mention that it's not always blue, but when you're near the Earth, it's visible. Oh, and maybe compare it to how rainbows are formed with different colors. So, putting it all together in a clear, concise manner.
</think>

The sky appears blue because of a process called **refraction**—where sunlight enters the Earth's atmosphere. Here's a breakdown:

1. **Light Refraction**: When sunlight enters the atmosphere, it interacts with gases like nitrogen and oxygen, which cause the light to bend.  
2. **Scattering by Gases**: These gases scatter the light in all directions, creating a "blue spectrum."  
3. **Earth's Atmosphere Difference**: The Earth's atmosphere is much thinner than the layers above the ocean (where clouds form), so the sky’s color is visible in the air without a sunscreen.  

In short: The sky appears blue because of refraction and scattering, resulting in the "blue light" we see."
`;

exports[`generateText > basic 2`] = `
[
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"content":"Why is the sky blue?","role":"user"}]",
    "ai.response.finishReason": "stop",
    "ai.response.id": "ae7361ee-4864-458d-bbcd-cebd88ba4860",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user is asking, "Why is the s",
    "ai.response.timestamp": "2025-08-30T15:53:28.359Z",
    "ai.usage.completionTokens": 472,
    "ai.usage.promptTokens": 14,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "196716a1-6ac8-477c-8cb6-ab246f6ee9bf",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "xsai",
    "gen_ai.usage.input_tokens": 14,
    "gen_ai.usage.output_tokens": 472,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "xsai",
    "ai.operationId": "ai.generateText",
    "ai.prompt": "{"messages":[{"content":"Why is the sky blue?","role":"user"}]}",
    "ai.response.finishReason": "stop",
    "ai.response.providerMetadata": "{}",
    "ai.response.text": "<think> Okay, so the user is asking, "Why is the s",
    "ai.usage.completionTokens": 472,
    "ai.usage.promptTokens": 14,
    "operation.name": "ai.generateText",
  },
]
`;

exports[`generateText > basic/ai 1`] = `"The sky is blue because it contains a specific type of **water vapor** that reflects a **part of the visible light** in the **blue spectrum**. This phenomenon is called the **dispersion of light** or **scattering of light** by the atmosphere. When sunlight enters the Earth's atmosphere, the water vapor in the air refracts the sunlight, causing it to spread out in a way that makes the sky appear blue. The blue color is the result of this scattering process, which is why the sky appears blue in the visible light."`;

exports[`generateText > basic/ai 2`] = `
[
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.generateText.doGenerate",
    "ai.prompt.messages": "[{"role":"user","content":[{"type":"text","text":"Why is the sky blue?"}]}]",
    "ai.response.finishReason": "stop",
    "ai.response.id": "aitxt-SvhhXW7TsBHuy5nUdYoR9QCN",
    "ai.response.model": "qwen3:0.6b",
    "ai.response.providerMetadata": "{"ollama":{}}",
    "ai.response.text": "The sky is blue because it contains a specific typ",
    "ai.response.timestamp": "2025-08-30T15:53:31.310Z",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.completionTokens": 112,
    "ai.usage.promptTokens": 22,
    "gen_ai.request.model": "qwen3:0.6b",
    "gen_ai.response.finish_reasons": [
      "stop",
    ],
    "gen_ai.response.id": "aitxt-SvhhXW7TsBHuy5nUdYoR9QCN",
    "gen_ai.response.model": "qwen3:0.6b",
    "gen_ai.system": "ollama.responses",
    "gen_ai.usage.input_tokens": 22,
    "gen_ai.usage.output_tokens": 112,
    "operation.name": "ai.generateText.doGenerate",
  },
  {
    "ai.model.id": "qwen3:0.6b",
    "ai.model.provider": "ollama.responses",
    "ai.operationId": "ai.generateText",
    "ai.prompt": "{"prompt":"Why is the sky blue?"}",
    "ai.response.finishReason": "stop",
    "ai.response.providerMetadata": "{"ollama":{}}",
    "ai.response.text": "The sky is blue because it contains a specific typ",
    "ai.settings.maxRetries": 2,
    "ai.settings.seed": 114514,
    "ai.usage.completionTokens": 112,
    "ai.usage.promptTokens": 22,
    "operation.name": "ai.generateText",
  },
]
`;
